src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs

```diff
@@ -1119,12 +1119,6 @@ public static bool UnsafeQueueUserWorkItem<TState>(Action<TState> callBack, TSta
             // internally we call UnsafeQueueUserWorkItemInternal directly for Tasks.
             if (ReferenceEquals(callBack, ThreadPool.s_invokeAsyncStateMachineBox))
             {
-                if (!(state is IAsyncStateMachineBox))
-                {
-                    // The provided state must be the internal IAsyncStateMachineBox (Task) type
-                    ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.state);
-                }
```

Comment:
I don't understand how this is safe.  With misuse, state could end up being any object, not an IAsyncStateMachineBox, not an IThreadPoolWorkItem, not a Task.  In which case aren't we going to do an invalid/unsafe/potentially-corrupting cast in https://github.com/dotnet/runtime/blob/45e20479eedff3aa3f19fed9b807cf3a189e8a95/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs#L610-L645?

Reply:
And then it will throw `EntryPointNotFoundException` when trying to wire up the `IThreadPoolWorkItem.Execute` interface call? The main issue is could it be corrupting; which would be a problem.

As for invalid, without this optimization it would try to do the same cast in the callback (though try to use the result) and throw an unhandled exception on the ThreadPool anyway

Reply:
> And then it will throw EntryPointNotFoundException when trying to wire up the IThreadPoolWorkItem.Execute interface call? 

Who knows what it would do: it's interpreting an object as something it's not, breaking type safety.

Reply:
This change is contentious in two ways... 😉